src/Cosserat/mapping/BaseCosserat.inl

```diff
@@ -1,274 +1,250 @@
+        (1.0 - std::cos(curv_abs_x_n * theta)) / std::pow(theta, 2);
+    double scalar2 = (curv_abs_x_n * theta - std::sin(curv_abs_x_n * theta)) /
+                     std::pow(theta, 3);
+    _g_X = I4 + curv_abs_x_n * Xi_hat_n + scalar1 * Xi_hat_n * Xi_hat_n +
+           scalar2 * Xi_hat_n * Xi_hat_n * Xi_hat_n;
+  }
+
+  if (d_debug.getValue())
+    msg_info("BaseCosserat: ") << "matrix _g_X : " << _g_X;
+  type::Mat3x3 M;
+  _g_X.getsub(0, 0, M); // get the rotation matrix
+
+  // convert the rotation 3x3 matrix to a quaternion
+  sofa::type::Quat<Real> R;
+  R.fromMatrix(M);
+  g_X_n = Transform(type::Vec3(_g_X(0, 3), _g_X(1, 3), _g_X(2, 3)), R);
 }
 
-
-//Fill exponential vectors
+// Fill exponential vectors
 template <class TIn1, class TIn2, class TOut>
-void BaseCosserat<TIn1, TIn2, TOut>::update_ExponentialSE3(const In1VecCoord & inDeform){
-    //helper::ReadAccessor<Data<helper::vector<double>>> curv_abs_input = d_curv_abs_section;
-    helper::ReadAccessor<Data<type::vector<double>>> curv_abs_frames = d_curv_abs_frames;
-    //m_index_input = 0;
-
-    m_framesExponentialSE3Vectors.clear();
-    m_nodesExponentialSE3Vectors.clear();
-    m_nodesLogarithmeSE3Vectors.clear();
-    const unsigned int sz = curv_abs_frames.size();
-
-    //Compute exponential at frame points
-    for (size_t i = 0; i < sz; i++) {
-        Transform T ;
-
-        const type::Vec3 k = inDeform[m_indicesVectors[i]-1];
-        const double x = m_framesLengthVectors[i];
-        computeExponentialSE3(x,k,T);
-        m_framesExponentialSE3Vectors.push_back(T);
-
-        if(d_debug.getValue()){
-            msg_info("BaseCosserat:")<< "__________________________________________";
-            msg_info("BaseCosserat:")<< "x :"<< x << "; k :"<< k;
-            msg_info("BaseCosserat:")<< "m_framesExponentialSE3Vectors :"<< T;
-        }
+void BaseCosserat<TIn1, TIn2, TOut>::update_ExponentialSE3(
+    const In1VecCoord &inDeform) {
+  // helper::ReadAccessor<Data<helper::vector<double>>> curv_abs_input =
+  // d_curv_abs_section;
+  helper::ReadAccessor<Data<type::vector<double>>> curv_abs_frames =
+      d_curv_abs_frames;
+
+  m_framesExponentialSE3Vectors.clear();
+  m_nodesExponentialSE3Vectors.clear();
+  m_nodesLogarithmeSE3Vectors.clear();
+  const unsigned int sz = curv_abs_frames.size();
+
+  // Compute exponential at each frame point
+  for (size_t i = 0; i < sz; i++) {
+    Transform g_X_frame_i;
+
+    const Coord1 strain_n = inDeform[m_indicesVectors[i] -
+                                     1]; // Cosserat reduce coordinates (strain)
+                                         // the size varies from 1 to 6
+    //  The distance between the frame and the closest beam node toward the base
+    const SReal curv_abs_x =
+        m_framesLengthVectors[i]; // curv_abs_x = frame_curv_abs - L_(n-1)
+    computeExponentialSE3(curv_abs_x, strain_n, g_X_frame_i);
+    m_framesExponentialSE3Vectors.push_back(g_X_frame_i);
+
+    if (d_debug.getValue()) {
```

Comment:
For object inhering from BaseObject, don't use the msg_info("aStringHere").
```cpp
msg_info_when(d_debug.getValue()) 
        << "_________________" << i << "_________________________"  << msgendl 
        << "x :"  << curv_abs_x << "; strain :" << strain_n << msgendl
        << "m_framesExponentialSE3Vectors :" << g_X_frame_i;
```
The use of d_debug is also questionnable because msg_info() is actually already containing a conditionnal printing... like if(d_printLog.getValue() == true){ ... } ... so why not juste use it ?